process/dns_v1.go

```diff
@@ -0,0 +1,336 @@
+package process
+
+import (
+	"bytes"
+	"encoding/binary"
+	"math"
+
+	"github.com/DataDog/mmh3"
+)
+
+// DNS data is encoded as a very basic bucketed hash table.  There are three blocks, or buffers, of data:
+//
+//	The "name" block is all of the unique DNS names.  The length of the name is stored as a varint
+//	followed by the name itself
+//
+//	The "bucket" block contains all of the hash buckets.  The format of each bucket is:
+//		varint for number of entries in bucket
+//		For each entry in the bucket:
+//			varint for length of ip
+//			ip bytes
+//			varint for number of names associated with the ip
+//			Each associated name is encoded as a varint which is the position of the actual name string in the name block
+//
+//	The "position" block is a list of varints, one for each bucket, where each varint is a pointer to the start
+//	of the bucket in the bucket block
+//
+// The overall buffer is encoded as:
+//	1 byte indicating version
+// 	1 byte indicating the number of buckets
+//	varint indicating the length of the name buffer.
+//	varint indicating the length of the position buffer
+// 	varint indicating the position of the "middle" (bucketCount / 2) bucket in the position block
+//		We will use this to skip the half of the buckets when searching for the target bucket index
+//	position block
+//	bucket block
+//	name block
+//
+// Notes:
+//	Using varints saves space at the cost of not having random access to certain sections of data, particularly the
+//	bucket position mapping.  This was a deliberate trade off to reduce the size of the payload and thus memory usage
+//
+//	Varints are also more finicky to deal with in terms of calculating required space ahead of time.  This increases
+//	the implementation complexity, or at least the line count, but we reduce allocations & memory usage by
+// 	pre-sizing the output buffers
+//
+// This type is not thread safe
+type V1DNSEncoder struct {
+	BucketFactor float64
+	scratch      [binary.MaxVarintLen64]byte // Used for varint encoding
+}
+
+type bucketEntry struct {
+	keys []string
+	size int
+}
+
+// 1 byte for version, 1 byte for bucket count
+const dns1Version1PreambleLength = 2
+
+// Used for calculating the number of buckets for a given input map.
+// Currently the bucket count is calculated as `len(input) * bucketFactor`
+const defaultBucketFactor = 0.75
+
+func NewV1DNSEncoder() DNSEncoder {
+	return &V1DNSEncoder{
+		BucketFactor: defaultBucketFactor,
+	}
+}
+
+func (e *V1DNSEncoder) Encode(dns map[string]*DNSEntry) []byte {
```

Comment:
should this take a writer or a []byte so the agent can re-use the buffer between check runs ?

Reply:
Yes, we could do so.  I'd prefer to wait until we see that it shows up as problematic in the profiler though.